# **Proyecto Integrador: Análisis de conjunto de datos de anuncios de venta de coches**

## **Descripción del Proyecto**
Realizaremos un anális exploratorio de datos EDA, para entender la información del conjunto de datos y justificar los gráficos que posteriormente llevaremos a un Dashboard interactivo.

## **INICIALIZACIÓN**

In [29]:
 # Se cargan librerÍas.

import pandas as pd
import plotly.express as px

## **CARGA DE DATOS**

In [30]:
# Se carga el Dataframe.

df = pd.read_csv('vehicles_us.csv')

## **PREPARAR DATOS**

In [31]:
# Vemos información general del DF.

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  str    
 3   condition     51525 non-null  str    
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  str    
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  str    
 8   type          51525 non-null  str    
 9   paint_color   42258 non-null  str    
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  str    
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), str(7)
memory usage: 7.7 MB


In [32]:
# Imprimimos una muestra del DF.
df.head(10)

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
0,9400,2011.0,bmw x5,good,6.0,gas,145000.0,automatic,SUV,NaN,1.0,2018-06-23,19
1,25500,NaN,ford f-150,good,6.0,gas,88705.0,automatic,pickup,white,1.0,2018-10-19,50
2,5500,2013.0,hyundai sonata,like new,4.0,gas,110000.0,automatic,sedan,red,NaN,2019-02-07,79
3,1500,2003.0,ford f-150,fair,8.0,gas,NaN,automatic,pickup,NaN,NaN,2019-03-22,9
4,14900,2017.0,chrysler 200,excellent,4.0,gas,80903.0,automatic,sedan,black,NaN,2019-04-02,28
5,14990,2014.0,chrysler 300,excellent,6.0,gas,57954.0,automatic,sedan,black,1.0,2018-06-20,15
6,12990,2015.0,toyota camry,excellent,4.0,gas,79212.0,automatic,sedan,white,NaN,2018-12-27,73
7,15990,2013.0,honda pilot,excellent,6.0,gas,109473.0,automatic,SUV,black,1.0,2019-01-07,68
8,11500,2012.0,kia sorento,excellent,4.0,gas,104174.0,automatic,SUV,NaN,1.0,2018-07-16,19
9,9200,2008.0,honda pilot,excellent,NaN,gas,147191.0,automatic,SUV,blue,1.0,2019-02-15,17


In [33]:
# Ver recuento y porcentaje de nulos por columna
nulos = pd.DataFrame({
    'nulos': df.isna().sum(),
    'porcentaje': (df.isna().sum() / len(df)) * 100
})
print(nulos[nulos['nulos'] > 0])

             nulos  porcentaje
model_year    3619    7.023775
cylinders     5260   10.208637
odometer      7892   15.316836
paint_color   9267   17.985444
is_4wd       25953   50.369723


In [34]:
# Buscamos duplicados implícitos en columna "model".
df['model'].dropna().unique()

<ArrowStringArray>
[                           'bmw x5',                        'ford f-150',
                    'hyundai sonata',                      'chrysler 200',
                      'chrysler 300',                      'toyota camry',
                       'honda pilot',                       'kia sorento',
          'chevrolet silverado 1500',                      'honda accord',
                          'ram 1500',                         'gmc yukon',
                     'jeep cherokee',                'chevrolet traverse',
                   'hyundai elantra',                   'chevrolet tahoe',
                       'toyota rav4',               'chevrolet silverado',
                     'jeep wrangler',                  'chevrolet malibu',
                    'ford fusion se',                  'chevrolet impala',
                'chevrolet corvette',                      'jeep liberty',
                   'toyota camry le',                     'nissan altima',
      

In [35]:
# Buscamos duplicados implícitos en columna "type".
df['type'].dropna().unique()

<ArrowStringArray>
[        'SUV',      'pickup',       'sedan',       'truck',       'coupe',
         'van', 'convertible',   'hatchback',       'wagon',    'mini-van',
       'other',     'offroad',         'bus']
Length: 13, dtype: str

In [36]:
# Contar filas duplicadas explícitas
duplicados = df.duplicated().sum()
print(f"Cantidad de registros duplicados: {duplicados}")

Cantidad de registros duplicados: 0


In [37]:
# Código actualizado para silenciar la advertencia
for col in df.select_dtypes(include=['object', 'str']).columns:
    coincidencias = df[col].astype(str).str.lower().eq('tbd').sum()
    print(f"Columna '{col}' con 'TBD': {coincidencias}")

Columna 'model' con 'TBD': 0
Columna 'condition' con 'TBD': 0
Columna 'fuel' con 'TBD': 0
Columna 'transmission' con 'TBD': 0
Columna 'type' con 'TBD': 0
Columna 'paint_color' con 'TBD': 0
Columna 'date_posted' con 'TBD': 0


In [38]:
# Resumen estadístico de las variables numéricas
df.describe()

,price,model_year,cylinders,odometer,is_4wd,days_listed
count,51525.000000,47906.000000,46265.000000,43633.000000,25572.0,51525.00000
mean,12132.464920,2009.750470,6.125235,115553.461738,1.0,39.55476
std,10040.803015,6.282065,1.660360,65094.611341,0.0,28.20427
min,1.000000,1908.000000,3.000000,0.000000,1.0,0.00000
25%,5000.000000,2006.000000,4.000000,70000.000000,1.0,19.00000
50%,9000.000000,2011.000000,6.000000,113000.000000,1.0,33.00000
75%,16839.000000,2014.000000,8.000000,155000.000000,1.0,53.00000
max,375000.000000,2019.000000,12.000000,990000.000000,1.0,271.00000


In [39]:
# Conocer cuántos vehículos tienen un precio atípico (menor a $100)
precios_bajos = (df['price'] < 100).sum()
print(f"Vehículos con precio menor a $100: {precios_bajos}")

Vehículos con precio menor a $100: 863


In [40]:
# Conocer cuántos vehículos se vendieron en 0 días
ventas_dia_cero = (df['days_listed'] == 0).sum()
print(f"Vehículos vendidos el mismo día (0 días): {ventas_dia_cero}")

Vehículos vendidos el mismo día (0 días): 54


### Diagnóstico y Plan de Tratamiento de Valores Ausentes

Tras inspeccionar las 51,525 filas del dataset, se identificaron valores nulos en 5 columnas:

1. **is_4wd (50.37%):** Los nulos representan la ausencia de tracción 4x4; se imputan con `0.0`.
2. **paint_color (17.99%):** Al ser categórica, se imputan con `'unknown'` para no perder información de otras variables.
3. **odometer (15.32%) y cylinders (10.21%):** Se imputan utilizando la mediana de cada variable agrupada por el modelo del vehículo.
4. **model_year (7.02%):** Se imputa con la mediana por modelo para preservar la totalidad de los datos.

### Hallazgos de Calidad de Datos y Casos Atípicos

1. **Inconsistencias en Precios:** Se detectaron 863 anuncios (1.67%) con precios inferiores a $100 (incluyendo valores de $1). Se establece la regla de filtrar `price >= 100` en las visualizaciones de la aplicación interactiva para mantener la precisión de la distribución.
2. **Velocidad de Venta (days_listed):** Un total de 54 vehículos (0.10%) se registran con 0 días de publicación, lo cual refleja ventas inmediatas o cancelaciones el mismo día de carga. Se conservan por ser registros legítimos del comportamiento del mercado.

## TRATAMIENTO Y CORRECCIÓN DE DATOS

In [41]:
# 1. Imputación de variables categóricas y booleanas
df['is_4wd'] = df['is_4wd'].fillna(0.0)
df['paint_color'] = df['paint_color'].fillna('unknown')

# 2. Imputación de numéricas utilizando la mediana agrupada por modelo de vehículo
df['model_year'] = df['model_year'].fillna(df.groupby('model')['model_year'].transform('median'))
df['cylinders'] = df['cylinders'].fillna(df.groupby('model')['cylinders'].transform('median'))
df['odometer'] = df['odometer'].fillna(df.groupby('model')['odometer'].transform('median'))

# 3. Filtrado de publicaciones con precios atípicos (inconsistentes)
df_cleaned = df[df['price'] >= 100].copy()

# 4. Verificación final de valores nulos en el DataFrame limpio
print("--- Conteo de nulos tras la limpieza ---")
print(df_cleaned.isna().sum())
print(f"\nTotal de filas resultantes: {len(df_cleaned)}")

--- Conteo de nulos tras la limpieza ---
price            0
model_year       0
model            0
condition        0
cylinders        0
fuel             0
odometer        41
transmission     0
type             0
paint_color      0
is_4wd           0
date_posted      0
days_listed      0
dtype: int64

Total de filas resultantes: 50662


In [42]:
# Imputación de la variable 'odometer' con la mediana global
df_cleaned['odometer'] = df_cleaned['odometer'].fillna(df_cleaned['odometer'].median())

df_cleaned.info()

<class 'pandas.DataFrame'>
Index: 50662 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         50662 non-null  int64  
 1   model_year    50662 non-null  float64
 2   model         50662 non-null  str    
 3   condition     50662 non-null  str    
 4   cylinders     50662 non-null  float64
 5   fuel          50662 non-null  str    
 6   odometer      50662 non-null  float64
 7   transmission  50662 non-null  str    
 8   type          50662 non-null  str    
 9   paint_color   50662 non-null  str    
 10  is_4wd        50662 non-null  float64
 11  date_posted   50662 non-null  str    
 12  days_listed   50662 non-null  int64  
dtypes: float64(4), int64(2), str(7)
memory usage: 8.0 MB


## ANÁLISIS EXPLORATORIO DE DATOS (EDA)

## Histograma: Distribución de Precios

In [46]:
# Histograma para analizar la distribución del precio
fig_hist = px.histogram(
    df_cleaned, 
    x='price',
    nbins=50,
    title='Distribución del Precio de los Vehículos',
    labels={'price': 'Precio del vehículo ($)'}
)

# Cambiar la etiqueta del eje Y
fig_hist.update_layout(yaxis_title='Cantidad de anuncios')

fig_hist.show()

### Comentarios respecto a la Distribución del Precio (Histograma)


* Eje X (Precio del vehículo): Muestra los rangos de costo a los que los vendedores ofertan sus autos.
* Eje Y (Cantidad de anuncios): Indica cuántos anuncios o publicaciones existen en la plataforma para cada rango de precio.

**Hallazgos principales:**
  1. Concentración en precios bajos: La barra más alta se ubica al principio del gráfico. Esto significa que la gran mayoría de las publicaciones ofertan vehículos económicos (concentrados principalmente entre $100 y $15,000 dólares).
  2. Caída rápida de la oferta: Conforme el precio aumenta, la cantidad de anuncios disminuye de forma drástica. 
  3. Valores poco comunes: Hay muy pocos vehículos con precios superiores a $50,000 dólares y casos contados que superan los $100,000 (autos de lujo, de trabajo o piezas de colección)[cite: 1].

* Podríamos concluir que el mercado de autos usados en esta plataforma, está dominado por vehículos accesibles de gama media y baja, con muy pocas opciones de precios elevados.

# Gráfico de Dispersión: Precio vs. Kilometraje

In [ ]:
# Gráfico de dispersión para analizar la depreciación por kilometraje
fig_scatter = px.scatter(
    df_cleaned, 
    x='odometer', 
    y='price', 
    title='Relación entre Kilometraje y Precio'
)
fig_scatter.show()

### Comentarios rspecto a la Relación entre Kilometraje y Precio (Gráfico de Dispersión)

  * Cada punto azul representa **un vehículo individual** publicado en la plataforma[cite: 1, 2].
  * Eje X (Odometer / Kilometraje): Indica el uso que tiene el auto medido en millas[cite: 1, 3].
  * Eje Y (Price / Precio): Indica el valor en dólares al que se está vendiendo[cite: 1, 3].

* **Hallazgos principales:**
  1. Tendencia de depreciación (Caída de precio): Se observa una clara inclinación hacia abajo. A medida que los puntos se desplazan hacia la derecha (más millas recorridas), la altura en el eje Y baja (menor precio pedido).
  2. El núcleo del mercado: La mayor cantidad de puntos forma una nube densa en la esquina inferior izquierda. Esto nos dice que el mercado común está compuesto por autos con menos de 200,000 millas (`0.2M`) y precios por debajo de $50,000 dólares.
  3. Puntos atípicos (Casos especiales):
     * **Arriba (Precios muy altos):** Puntos aislados con precios cercanos a $300,000 o $375,000 dólares (posibles vehículos comerciales o errores de captura al publicar).
     * **A la extrema derecha (Kilometraje extremo):** Puntos solitarios cerca de 1,000,000 de millas (`1M`), que corresponden a autos con uso desmedido o números arbitrarios colocados por el vendedor.

* Podríoamos concluir que el desgaste por uso (kilometraje) es un factor determinante en el valor del automóvil: a mayor recorrido, menor es su costo en el mercado.

In [ ]:


car_data = pd.read_csv('vehicles_us.csv') # leer los datos
fig = px.histogram(car_data, x="odometer") # crear un histograma
fig.show() # crear gráfico de dispersión

In [ ]:
car_data = pd.read_csv('vehicles_us.csv') # leer los datos
fig = px.scatter(car_data, x="odometer", y="price") # crear un gráfico de dispersión
fig.show() # crear gráfico de dispersión